### ISSUE TITLE
Push to github fails

### ISSUE THREAD

**Issue by feinfinger** [tomopy/tomopy#239]
I tried to push a local commit to github but got rejected with an 403 error (from Linux using pycharm). Do I   need to be somehow registered in order to able to push my code to the repo? Here is the error:
Failed with error: unable to access 'https://github.com/tomopy/tomopy.git/': The requested URL returned error: 403
Switching to ssh doesn't seem to work either (could not read from remote repository).

**Comment by dgursoy** [comment]
You can't directly push your local changes to main repo, but you should be able to send pull requests. Here's some more explanation how to do that: http://tomopy.readthedocs.io/en/latest/devguide.html#contributing-back

**Comment by feinfinger** [comment]
Ok. I tried to follow the explanation and it says there:  'You will need to push your local commits to GitHub, then...' - if that does not mean the tomopy repo, which repo do I push to?
Sorry for asking these beginners questions, but I'm a bit confused by the explanation on the website and maybe that helps to improve the documentation...

**Comment by dmpelt** [comment]
Typically, you first [fork](https://help.github.com/articles/fork-a-repo/) the repository on the GitHub site, which will create a copy of the git repository in your personal account (so in your case, will create https://github.com/feinfinger/tomopy). You can push changes to this repository. After you are finished working on a certain feature, and think it is ready to be included in TomoPy, you can create a [Pull Request](https://help.github.com/articles/creating-a-pull-request/), after which the TomoPy maintainers can accept them into the main TomoPy code, or request further changes. This all works best when you develop in feature-specific branches (see [here](https://guides.github.com/introduction/flow/)).

Let me know if you run into any trouble: it looks a bit complicated at first, but it is the standard way of doing things with GitHub, and very useful to be able to collaborate in large projects.

**Comment by prjemian** [comment]
Your difficulty is that you don't have write (or push) access to the 
tomopy/tomopy repository.  The repo allows you to clone a copy to your 
local storage.  You can modify that locally as much as you wish, 
including making commits locally.  That is *your* repository on *your* 
own storage.  You just don't have permission to push that back to the 
tomopy/tomopy repo.

This is a common situation for repositories, certainly not unique to 
tomopy.  In fact, git and GitHub were designed for just this type of 
thing (workflow).  Here's what you do instead:

0. save your local modified tomopy directory in a safe place since you 
will create a new tomopy directory from github shortly

1. login to GitHub web interface with your github account

2. Using the GitHub web interface, go to the tomopy/tomopy repo main 
page:  https://github.com/tomopy/tomopy

3. on the right side of that page, the line just below your icon, click 
on the "Fork" text (not the number but the word "Fork")

4. For me, it asks "Where should we fork this repository?"  Maybe you 
get that question if you are a member of one or more projects.  Choose 
your own account (?feinfinger?).

5. When that fork is complete, go to your account and pick your fork of 
tomopy.

6. Clone it (https://github.com/feinfinger/tomopy.git) to your hard 
disk.  (This is the "new" copy described in step 0 above.)

7. bring the changes from your "safe" copy of tomopy (step 0 above) and 
commit

8. push that commit back to https://github.com/feinfinger/tomopy.git

9. make a "Pull request" (on https://github.com/tomopy/tomopy)
    that https://github.com/feinfinger/tomopy.git
    be merged into https://github.com/tomopy/tomopy.git

This will bring you to the condition you expect to be in today.  The 
tomopy owners will examine your PR and respond.

Pete

**Comment by prjemian** [comment]
And, as Daniel said, creating a feature-specific branch (after step 5 or step 6) is the *best* way.

You create an issue for the feature you want on tomopy/tomopy.
That gets assigned a number, say 512 for hypothetical syntax checking feature
You create a branch with a name that describes the feature in a word or-couple-words and include the issue number as well: syntax-check-512
After you clone your fork locally, you checkout that branch and make your changes there.
Push that back to GitHub and make a Pull Request from that branch.

**Comment by feinfinger** [comment]
Excellent, thanks everyone! I've got it finally. I'm not sure if it's worth mentioning that you need a local fork on the http://tomopy.readthedocs.io/en/latest/devguide.html#contributing-back page but maybe that is too common knowledge...
